Framework/src/BookkeepingQualitySink.cxx

```diff
@@ -0,0 +1,139 @@
+#include "QualityControl/QualitiesToFlagCollectionConverter.h"
+#include "QualityControl/QualityObject.h"
+#include "QualityControl/QcInfoLogger.h"
+#include <BookkeepingApi/QcFlagServiceClient.h>
+#include <BookkeepingApi/BkpClientFactory.h>
+#include <stdexcept>
+
+namespace o2::quality_control::core
+{
+
+void BookkeepingQualitySink::send(const std::string& grpcUri, const BookkeepingQualitySink::FlagsMap& flags, Provenance type)
+{
+  auto bkpClient = o2::bkp::api::BkpClientFactory::create(grpcUri);
+  auto& qcClient = bkpClient->qcFlag();
+
+  ILOG(Info, Support) << "Sending " << flags.size() << " flag collections" << ENDM;
+
+  for (const auto& [_, flagCollection] : flags) {
+
+    std::vector<QcFlag> bkpQcFlags{};
+    for (const auto& flag : *flagCollection) {
+      // BKP uses start/end of run for missing time values, so we are using this functionality in order to avoid
+      // determining these values by ourselves (see TaskRunner::start() for details). mtichak checked with mboulais that
+      // it is okay to do so.
+      bkpQcFlags.emplace_back(QcFlag{
+        .flagTypeId = flag.getFlag().getID(),
+        .from = flag.getStart() == gFullValidityInterval.getMin() ? std::nullopt : std::optional<uint64_t>{ flag.getStart() },
+        .to = flag.getEnd() == gFullValidityInterval.getMax() ? std::nullopt : std::optional<uint64_t>{ flag.getEnd() },
+        .origin = flag.getSource(),
+        .comment = flag.getComment() });
+    }
+
+    if (bkpQcFlags.empty()) {
+      continue;
+    }
+
+    try {
+      switch (type) {
+        case Provenance::SyncQC:
+        case Provenance::AsyncQC:
+          qcClient->createForDataPass(flagCollection->getRunNumber(), flagCollection->getPassName(), flagCollection->getDetector(), bkpQcFlags);
+          break;
+        case Provenance::MCQC:
+          qcClient->createForSimulationPass(flagCollection->getRunNumber(), flagCollection->getPeriodName(), flagCollection->getDetector(), bkpQcFlags);
+          break;
+      }
+    } catch (const std::runtime_error& err) {
+      ILOG(Error, Support) << "Failed to send flags for detector: " << flagCollection->getDetector()
+                           << " and pass: " << flagCollection->getPassName()
+                           << " with error: " << err.what() << ENDM;
+    }
+  }
+}
+
+BookkeepingQualitySink::BookkeepingQualitySink(const std::string& grpcUri, Provenance type, SendCallback sendCallback)
+  : mGrpcUri{ grpcUri }, mType{ type }, mSendCallback{ sendCallback } {}
+
+auto merge(std::unique_ptr<QualityControlFlagCollection>&& collection, const std::unique_ptr<QualityObject>& qualityObject) -> std::unique_ptr<QualityControlFlagCollection>
+{
+  QualitiesToFlagCollectionConverter converter(std::move(collection), qualityObject->getPath());
+  converter(*qualityObject);
+  return converter.getResult();
+}
+
+auto collectionFromQualityObject(const QualityObject& qualityObject) -> std::unique_ptr<QualityControlFlagCollection>
```

Comment:
Given it's name, it's slightly unclear to me that this helper creates the collection without putting the QO itself in it.

Perhaps it could be renamed to `collectionForQualityObject`?